### PR TITLE
[SPARK-20094][SQL] Preventing push down of IN subquery to Join operator

### PR DESCRIPTION
**Pull request by wzhfy**
## What changes were proposed in this pull request?

TPCDS q45 fails becuase:
`ReorderJoin` collects all predicates and try to put them into join condition when creating ordered join. If a predicate with an IN subquery (`ListQuery`) is in a join condition instead of a filter condition, `RewritePredicateSubquery.rewriteExistentialExpr` would fail to convert the subquery to an `ExistenceJoin`, and thus result in error.

We should prevent push down of IN subquery to Join operator.

## How was this patch tested?

Add a new test case in `FilterPushdownSuite`.

### CODE REVIEW

**Comment by wzhfy** [comment]
cc @nsyca @hvanhovell

**Comment by wzhfy** [comment]
also cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 1]
```diff
@@ -90,6 +90,7 @@ trait PredicateHelper {
    * Returns true iff `expr` could be evaluated as a condition within join.
    */
   protected def canEvaluateWithinJoin(expr: Expression): Boolean = expr match {
+    case l: ListQuery => false
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: Doesn't the following `case e: SubqueryExpression` prevent it?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: add some comments to explain why `ListQuery` should always return false even the children is empty.

Comment (wzhfy) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: ok

Comment (wzhfy) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: @viirya Now pulling up IN predicates is deferred to `RewritePredicateSubquery`. As a result, `ListQuery`'s children can be empty before that rule and falls into `case e: SubqueryExpression`.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: OK. The comment looks good.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 2]
```diff
@@ -90,6 +90,10 @@ trait PredicateHelper {
    * Returns true iff `expr` could be evaluated as a condition within join.
    */
   protected def canEvaluateWithinJoin(expr: Expression): Boolean = expr match {
+    case l: ListQuery =>
+      // Now pulling up IN predicates is deferred to `RewritePredicateSubquery`. As a result,
+      // `ListQuery`'s children can be empty before that rule and falls into the next case.
+      false
```
Comment (nsyca) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: Suggested text in the comment:

[NOT] IN (<ListQuery>) expression cannot be evaluated as part of a Join operator. Currently the only way to evaluate a [NOT] IN subquery is to convert it to a LeftSemi or LeftAnti by `RewritePredicateSubquery` rule.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/JoinOptimizationSuite.scala [hunk 3]
```diff
@@ -123,6 +136,31 @@ class JoinOptimizationSuite extends PlanTest {
...
+
+    val queryPlan = x.join(y).join(z)
+      .where(("x.b".attr === "z.b".attr) && ("y.d".attr === "z.a".attr) &&
+        ("x.a".attr > 1 || "z.c".attr.in(ListQuery(w.select("w.d".attr)))))
+
+    val expectedPlan = x.join(z, Inner, Some("x.b".attr === "z.b".attr))
+      .join(w, ExistenceJoin(exists), Some("z.c".attr === "w.d".attr))
+      .where("x.a".attr > 1 || exists)
+      .select("x.a".attr, "x.b".attr, "x.c".attr, "z.a".attr, "z.b".attr, "z.c".attr)
+      .join(y, Inner, Some("y.d".attr === "z.a".attr))
```
Comment (nsyca) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/JoinOptimizationSuite.scala]: This is a side question. It's not a review comment.

Is there any mechanism in this test suite that ensures that the plan Join( Join(x, z), y) is picked by the join reorder algorithm? Isn't a Join( Join(y, z), x) is also a viable plan if the join predicate (y.d = z.a) is estimated to be more selective than the predicate sets between x and z?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala [hunk 4]
```diff
@@ -289,6 +289,9 @@ abstract class QueryPlan[PlanType <: QueryPlan[PlanType]] extends TreeNode[PlanT
       case m: Map[_, _] => m
       case d: DataType => d // Avoid unpacking Structs
       case seq: Traversable[_] => seq.map(recursiveTransform)
+      case ExistenceJoin(exists: Attribute) =>
+        // `exists` must be an Attribute in ExistenceJoin
+        ExistenceJoin(transformExpression(exists).asInstanceOf[Attribute])
```
Comment (wzhfy) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala]: Because `ExistenceJoin` only accepts `Attribute` as its parameter type.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala (hunk 2 above)
Comment (wzhfy) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: ok, thanks!

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/JoinOptimizationSuite.scala (hunk 3 above)
Comment (wzhfy) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/JoinOptimizationSuite.scala]: This test suite is for `JoinReorder` rule. This rule is not cost based and it's not related to estimation. It only searches for joinable tables and put cartesian product to the end of the plan.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala (hunk 4 above)
Comment (nsyca) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala]: Can you give an example of `transformExpression` of an instance of `Attribute` not returning a value of type `Attribute` or its subtypes?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/JoinOptimizationSuite.scala (hunk 3 above)
Comment (nsyca) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/JoinOptimizationSuite.scala]: My question was why Optimizer does not pick a plan of Join( Join(y, z), x).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala (hunk 4 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala]: Why is this change necessary? This smells for two reasons:

- `QueryPlan` shouldn't have know about such a detail. This means something is wrong with the abstraction.
- Why do we want to change the attribute in an ExistenceJoin? It is produced by the join, and we really should not replace it.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/JoinOptimizationSuite.scala (hunk 3 above)
Comment (nsyca) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/JoinOptimizationSuite.scala]: I guess your test case depends on the fact that currently we don't have an optimization to push down table y below the join of (x, z).

Essentially, this is not a problem specifically to `ReorderJoin` as implied from the title of the PR. The problem is at the definition `canEvaluateWithinJoin`. I can demonstrate the same problem without the table y in this test case.

A slight modification of your test case as shown below also reveals the problem.

````
    val queryPlan = x.join(z)
      .where(("x.b".attr === "z.b".attr) &&
        ("x.a".attr > 1 || "z.c".attr.in(ListQuery(w.select("w.d".attr)))))

    val expectedPlan = x.join(z, Inner, Some("x.b".attr === "z.b".attr))
      .join(w, ExistenceJoin(exists), Some("z.c".attr === "w.d".attr))
      .where("x.a".attr > 1 || exists)
      .select("x.a".attr, "x.b".attr, "x.c".attr, "z.a".attr, "z.b".attr, "z.c".attr)
````

which has a call to `canEvaluateWithinJoin` from the rule `PushPredicateThroughJoin`.

Comment (nsyca) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/JoinOptimizationSuite.scala]: May I suggest the following:
- Change the title of the PR to "Preventing push down of IN subquery to Join operator"
- Update the test case to its bare minimum by removing the join to table y
- Change the title of the test case to reflect the new test

On the last point, removing the "reorder inner joins - " should do.

Comment (wzhfy) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/JoinOptimizationSuite.scala]: Yea, thanks for the suggestion. Also I will remove the test case to `FilterPushdownSuite` since now it will not go into `ReorderJoin` rule.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala (hunk 4 above)
Comment (wzhfy) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala]: Yes I feel the same. But after modification as suggested by @nsyca , this change can also be removed. Because I move the test case to `FilterPushdownSuite`, we don't have to touch the `RewritePredicateSubquery` rule.

**Comment by hvanhovell** [comment]
LGTM - pending jenkins.

**Comment by hvanhovell** [comment]
Merging to master. Thanks!
